metadata/src/retain.rs

```diff
@@ -0,0 +1,231 @@
+fn update_type(ty: &mut UntrackedSymbol<TypeId>, map_ids: &BTreeMap<u32, u32>) {
+    let old_id = ty.id;
+    let new_id = map_ids
+        .get(&old_id)
+        .unwrap_or_else(|| panic!("PortableRegistry did not retain type id {old_id}. This is a bug. Please open an issue."));
+    *ty = (*new_id).into();
+}
+
+/// Update all type IDs of the provided pallet using the new type IDs from the portable registry.
+fn update_pallet_types(pallet: &mut PalletMetadata<PortableForm>, map_ids: &BTreeMap<u32, u32>) {
+    if let Some(storage) = &mut pallet.storage {
+        for entry in &mut storage.entries {
+            match &mut entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    update_type(ty, map_ids);
+                }
+                StorageEntryType::Map { key, value, .. } => {
+                    update_type(key, map_ids);
+                    update_type(value, map_ids);
+                }
+            }
+        }
+    }
+
+    if let Some(calls) = &mut pallet.calls {
+        update_type(&mut calls.ty, map_ids);
+    }
+
+    if let Some(event) = &mut pallet.event {
+        update_type(&mut event.ty, map_ids);
+    }
+
+    for constant in &mut pallet.constants {
+        update_type(&mut constant.ty, map_ids);
+    }
+
+    if let Some(error) = &mut pallet.error {
+        update_type(&mut error.ty, map_ids);
+    }
+}
+
+/// Update all type IDs of the provided extrinsic metadata using the new type IDs from the portable registry.
+fn update_extrinsic_types(
+    extrinsic: &mut ExtrinsicMetadata<PortableForm>,
+    map_ids: &BTreeMap<u32, u32>,
+) {
+    update_type(&mut extrinsic.ty, map_ids);
+
+    for signed in &mut extrinsic.signed_extensions {
+        update_type(&mut signed.ty, map_ids);
+        update_type(&mut signed.additional_signed, map_ids);
+    }
+}
+
+/// Generate a subset of the metadata that contains only the
+/// types needed to represent the provided pallets.
+///
+/// # Note
+///
+/// Used to strip metadata of unneeded information and to reduce the
+/// binary size.
+///
+/// # Panics
+///
+/// Panics if the [`scale_info::PortableRegistry`] did not retain all needed types,
+/// or the metadata does not contain the "sp_runtime::DispatchError" type.
+pub fn retain_metadata_pallets<F>(metadata: &mut RuntimeMetadataV14, mut filter: F)
+where
+    F: FnMut(&PalletMetadata<PortableForm>) -> bool,
+{
+    let mut type_ids = HashSet::new();
+
+    // Collect all types needed to represent the provided pallets.
+    metadata.pallets.retain(|pallet| {
+        if filter(pallet) {
+            collect_pallet_types(pallet, &mut type_ids);
+            true
+        } else {
+            false
+        }
+    });
+
+    let mut extrinsic_types = HashSet::new();
+    collect_extrinsic_types(&metadata.extrinsic, &mut extrinsic_types);
+    for id in extrinsic_types {
+        remove_generic_type_params(metadata, id);
```

Comment:
I find this these lines quite hard to read.

Not sure if it's just me that have troubles to understand that `remove_generics_type_params` just from the metadata itself.

Why do we need to insert that `id` in `type_ids`? That shouldn't exist anyway in the metadata?

Reply:
I wasn't a fan of this either so much, so I removed it and took a slightly different approach :)